src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -57,6 +57,44 @@
 #include "utilities/events.hpp"
 #include "utilities/preserveException.hpp"
 
+class CleanupObjectMonitorsHashtable: StackObj {
+ public:
+  bool do_entry(JavaThread*& key, ObjectMonitorsHashtable::PtrList*& list) {
+    list->clear();  // clear the LinkListNodes
+    delete list;    // then delete the LinkedList
+    return true;
+  }
+};
+
+ObjectMonitorsHashtable::~ObjectMonitorsHashtable() {
+  CleanupObjectMonitorsHashtable cleanup;
+  _ptrs->unlink(&cleanup);  // cleanup the LinkedLists
+  delete _ptrs;             // then delete the hash table
+}
+
+void ObjectMonitorsHashtable::add_entry(JavaThread* jt, ObjectMonitor* om) {
+  ObjectMonitorsHashtable::PtrList* list = get_entry(jt);
+  if (list != nullptr) {
+    // Add to existing list from the hash table:
+    list->add(om);
```

Comment:
These two lines are executed in both the if and else clause. Consider putting it after instead.